File: src/shared/melt.ts
// Núcleo (puro) da calculadora de derretimento/Alchemy (D5). O parser monta a lista de
// candidatos a partir do save (classifyItem + locais) e esta função aplica as regras de
// exclusão (equipado, Legendary+) e soma ouro de venda + XP de Cubo do catálogo meltData.ts.
import { meltOf } from './meltData'
import type { MeltRarityRow, MeltSummary } from './types'

// Um item de gear candidato ao derretimento, já com os dados que importam para a regra.
export interface MeltCandidate {
  key: number // ItemKey (para consultar valor no catálogo)
  gradeTier: number // tier da raridade (índice em GRADES)
  marketable: boolean // Legendary+ (vendável no Market → não derrete)
  equipped: boolean // equipado em algum herói → não derrete
}

/**
 * Soma o ouro de venda + XP de Cubo do gear derretível, excluindo equipados e Legendary+.
 * `byRarity` traz só as raridades com itens, em ordem crescente de tier.
 */
export function summarizeMelt(items: MeltCandidate[]): MeltSummary {
  const byTier = new Map<number, MeltRarityRow>()
  let totalGold = 0
  let totalCubeXp = 0
  let itemCount = 0
  let excludedMarketable = 0
  let excludedEquipped = 0
  let noData = 0

  for (const item of items) {
    if (item.equipped) {
      excludedEquipped++
      continue
    }
    if (item.marketable) {
      excludedMarketable++
      continue
    }
    const melt = meltOf(item.key)
    if (!melt) {
      noData++
      continue
    }
    const [gold, cubeXp] = melt
    totalGold += gold
    totalCubeXp += cubeXp
    itemCount++
    let row = byTier.get(item.gradeTier)
    if (!row) {
      row = { tier: item.gradeTier, count: 0, gold: 0, cubeXp: 0 }
      byTier.set(item.gradeTier, row)
    }
    row.count++
    row.gold += gold
    row.cubeXp += cubeXp
  }

  return {
    totalGold,
    totalCubeXp,
    itemCount,
    excludedMarketable,
    excludedEquipped,
    noData,
    byRarity: [...byTier.values()].sort((a, b) => a.tier - b.tier)
  }
}

File: dotnet/TbhTracker.Core/Logic/Melt.cs
using TbhTracker.Core.Models;

namespace TbhTracker.Core.Logic;

// Port de src/shared/melt.ts (D5). Nucleo puro da calculadora de derretimento/Alchemy:
// o parser monta os candidatos de gear e aqui aplicamos as regras de exclusao (equipado,
// Legendary+) e somamos ouro de venda + XP de Cubo do catalogo meltData.json.

/// <summary>Um item de gear candidato ao derretimento, com o que importa para a regra.</summary>
public sealed class MeltCandidate
{
    public int Key { get; init; }          // ItemKey (consulta no catalogo)
    public int GradeTier { get; init; }    // tier da raridade (indice em Items.Grades)
    public bool Marketable { get; init; }  // Legendary+ (vendavel no Market -> nao derrete)
    public bool Equipped { get; init; }    // equipado em algum heroi -> nao derrete
}

public static class Melt
{
    /// <summary>Ouro de venda + XP de Cubo de um ItemKey (null se fora do catalogo).</summary>
    public static (int Gold, int CubeXp)? MeltOf(int key)
    {
        if (Catalog.Melt.Items.TryGetValue(key.ToString(), out var v) && v.Length >= 2)
            return (v[0], v[1]);
        return null;
    }

    /// <summary>Soma o ouro de venda + XP de Cubo do gear derretivel, excluindo equipados e
    /// Legendary+. ByRarity traz so as raridades com itens, em ordem crescente de tier.</summary>
    public static MeltSummary Summarize(IEnumerable<MeltCandidate> items)
    {
        var byTier = new Dictionary<int, MeltRarityRow>();
        double totalGold = 0;
        double totalCubeXp = 0;
        var itemCount = 0;
        var excludedMarketable = 0;
        var excludedEquipped = 0;
        var noData = 0;

        foreach (var item in items)
        {
            if (item.Equipped) { excludedEquipped++; continue; }
            if (item.Marketable) { excludedMarketable++; continue; }

            var melt = MeltOf(item.Key);
            if (melt == null) { noData++; continue; }

            var (gold, cubeXp) = melt.Value;
            totalGold += gold;
            totalCubeXp += cubeXp;
            itemCount++;

            if (!byTier.TryGetValue(item.GradeTier, out var row))
            {
                row = new MeltRarityRow { Tier = item.GradeTier };
                byTier[item.GradeTier] = row;
            }
            row.Count++;
            row.Gold += gold;
            row.CubeXp += cubeXp;
        }

        return new MeltSummary
        {
            TotalGold = totalGold,
            TotalCubeXp = totalCubeXp,
            ItemCount = itemCount,
            ExcludedMarketable = excludedMarketable,
            ExcludedEquipped = excludedEquipped,
            NoData = noData,
            ByRarity = byTier.Values.OrderBy(r => r.Tier).ToList()
        };
    }
}
